JDK 17 InaccessibleObjectException with infinispan

### Area

infinispan

### Describe the bug

Using JDK 17 with a Keycloak Cluster (read: infinispan) results in InaccessibleObjectException upon execution of certain kcadm commands, such as adding a new IdentityStorageProvider.

### Version

20.0.1

### Expected behavior

Executing kcadm commands should not fail due to InaccessibleObjectException.

### Actual behavior

Executing an kcadm add ldap storage provider command results in the following stack trace:

```
2022-11-21 22:24:53,211 ERROR [org.infinispan.CLUSTER] (jgroups-12,keycloak1-26964) ISPN000474: Error processing request 150@keycloak2-62845: java.lang.ExceptionInInitializerError
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1268)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1064)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1381)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
        at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:231)
        at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
        at org.infinispan.jboss.marshalling.commons.AbstractJBossMarshaller.objectFromObjectStream(AbstractJBossMarshaller.java:137)
        at org.infinispan.jboss.marshalling.commons.AbstractJBossMarshaller.objectFromByteBuffer(AbstractJBossMarshaller.java:115)
        at org.infinispan.commons.marshall.AbstractMarshaller.objectFromByteBuffer(AbstractMarshaller.java:82)
        at org.infinispan.marshall.core.impl.DelegatingUserMarshaller.objectFromByteBuffer(DelegatingUserMarshaller.java:63)
        at org.infinispan.marshall.protostream.impl.MarshallableUserObject$Marshaller.read(MarshallableUserObject.java:139)
        at org.infinispan.marshall.protostream.impl.MarshallableUserObject$Marshaller.read(MarshallableUserObject.java:100)
        at org.infinispan.protostream.impl.ProtobufTagMarshallerDelegate.unmarshall(ProtobufTagMarshallerDelegate.java:32)
        at org.infinispan.protostream.WrappedMessage.readMessage(WrappedMessage.java:539)
        at org.infinispan.protostream.WrappedMessage.read(WrappedMessage.java:351)
        at org.infinispan.protostream.ProtobufUtil.fromWrappedByteArray(ProtobufUtil.java:129)
        at org.infinispan.marshall.protostream.impl.AbstractInternalProtoStreamMarshaller.objectFromByteBuffer(AbstractInternalProtoStreamMarshaller.java:97)
        at org.infinispan.marshall.protostream.impl.AbstractInternalProtoStreamMarshaller.objectFromByteBuffer(AbstractInternalProtoStreamMarshaller.java:92)
        at org.infinispan.marshall.core.GlobalMarshaller.readUnknown(GlobalMarshaller.java:905)
        at org.infinispan.marshall.core.GlobalMarshaller.readUnknown(GlobalMarshaller.java:890)
        at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:714)
        at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
        at org.infinispan.marshall.core.BytesObjectInput.readObject(BytesObjectInput.java:32)
        at org.keycloak.cluster.infinispan.WrapperClusterEvent$ExternalizerImpl.readObjectVersion1(WrapperClusterEvent.java:145)
        at org.keycloak.cluster.infinispan.WrapperClusterEvent$ExternalizerImpl.readObject(WrapperClusterEvent.java:130)
        at org.keycloak.cluster.infinispan.WrapperClusterEvent$ExternalizerImpl.readObject(WrapperClusterEvent.java:109)
        at org.infinispan.marshall.core.GlobalMarshaller.readAnnotated(GlobalMarshaller.java:735)
        at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:712)
        at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
        at org.infinispan.marshall.core.BytesObjectInput.readObject(BytesObjectInput.java:32)
        at org.infinispan.commands.write.PutKeyValueCommand.readFrom(PutKeyValueCommand.java:96)
        at org.infinispan.marshall.exts.ReplicableCommandExternalizer.readCommandParameters(ReplicableCommandExternalizer.java:117)
        at org.infinispan.marshall.exts.ReplicableCommandExternalizer.readObject(ReplicableCommandExternalizer.java:106)
        at org.infinispan.marshall.exts.ReplicableCommandExternalizer.readObject(ReplicableCommandExternalizer.java:68)
        at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
        at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
        at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
        at org.infinispan.marshall.core.BytesObjectInput.readObject(BytesObjectInput.java:32)
        at org.infinispan.commands.remote.SingleRpcCommand.readFrom(SingleRpcCommand.java:56)
        at org.infinispan.marshall.exts.ReplicableCommandExternalizer.readCommandParameters(ReplicableCommandExternalizer.java:117)
        at org.infinispan.marshall.exts.CacheRpcCommandExternalizer.readObject(CacheRpcCommandExternalizer.java:170)
        at org.infinispan.marshall.exts.CacheRpcCommandExternalizer.readObject(CacheRpcCommandExternalizer.java:88)
        at org.infinispan.marshall.core.GlobalMarshaller.readWithExternalizer(GlobalMarshaller.java:727)
        at org.infinispan.marshall.core.GlobalMarshaller.readNonNullableObject(GlobalMarshaller.java:708)
        at org.infinispan.marshall.core.GlobalMarshaller.readNullableObject(GlobalMarshaller.java:357)
        at org.infinispan.marshall.core.GlobalMarshaller.objectFromObjectInput(GlobalMarshaller.java:191)
        at org.infinispan.marshall.core.GlobalMarshaller.objectFromByteBuffer(GlobalMarshaller.java:220)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processRequest(JGroupsTransport.java:1455)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.processMessage(JGroupsTransport.java:1395)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport.access$300(JGroupsTransport.java:146)
        at org.infinispan.remoting.transport.jgroups.JGroupsTransport$ChannelCallbacks.up(JGroupsTransport.java:1586)
        at org.jgroups.JChannel.up(JChannel.java:780)
        at org.jgroups.stack.ProtocolStack.up(ProtocolStack.java:913)
        at org.jgroups.protocols.FRAG3.up(FRAG3.java:165)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:363)
        at org.jgroups.protocols.FlowControl.up(FlowControl.java:355)
        at org.jgroups.protocols.pbcast.GMS.up(GMS.java:876)
        at org.jgroups.protocols.pbcast.STABLE.up(STABLE.java:254)
        at org.jgroups.protocols.UNICAST3.up(UNICAST3.java:392)
        at org.jgroups.protocols.pbcast.NAKACK2.deliver(NAKACK2.java:926)
        at org.jgroups.protocols.pbcast.NAKACK2.handleMessage(NAKACK2.java:818)
        at org.jgroups.protocols.pbcast.NAKACK2.up(NAKACK2.java:602)
        at org.jgroups.protocols.VERIFY_SUSPECT.up(VERIFY_SUSPECT.java:132)
        at org.jgroups.protocols.FailureDetection.up(FailureDetection.java:186)
        at org.jgroups.protocols.FD_SOCK.up(FD_SOCK.java:254)
        at org.jgroups.protocols.MERGE3.up(MERGE3.java:281)
        at org.jgroups.protocols.Discovery.up(Discovery.java:300)
        at org.jgroups.protocols.TP.passMessageUp(TP.java:1400)
        at org.jgroups.util.SubmitToThreadPool$SingleMessageHandler.run(SubmitToThreadPool.java:98)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected java.util.AbstractMap() accessible: module java.base does not "opens java.util" to unnamed module @396a51ab
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
        at java.base/java.lang.reflect.Constructor.checkCanSetAccessible(Constructor.java:188)
        at java.base/java.lang.reflect.Constructor.setAccessible(Constructor.java:181)
        at org.jboss.marshalling.reflect.JDKSpecific$SerMethods.<init>(JDKSpecific.java:147)
        at org.jboss.marshalling.reflect.SerializableClass.<init>(SerializableClass.java:84)
        at org.jboss.marshalling.reflect.SerializableClassRegistry$1.computeValue(SerializableClassRegistry.java:62)
        at org.jboss.marshalling.reflect.SerializableClassRegistry$1.computeValue(SerializableClassRegistry.java:59)
        at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:228)
        at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:210)
        at java.base/java.lang.ClassValue.get(ClassValue.java:116)
        at org.jboss.marshalling.reflect.SerializableClassRegistry.lookup(SerializableClassRegistry.java:83)
        at org.jboss.marshalling.reflect.SerializableClass.<init>(SerializableClass.java:76)
        at org.jboss.marshalling.reflect.SerializableClassRegistry$1.computeValue(SerializableClassRegistry.java:62)
        at org.jboss.marshalling.reflect.SerializableClassRegistry$1.computeValue(SerializableClassRegistry.java:59)
        at java.base/java.lang.ClassValue.getFromHashMap(ClassValue.java:228)
        at java.base/java.lang.ClassValue.getFromBackup(ClassValue.java:210)
        at java.base/java.lang.ClassValue.get(ClassValue.java:116)
        at org.jboss.marshalling.reflect.SerializableClassRegistry.lookup(SerializableClassRegistry.java:83)
        at org.jboss.marshalling.river.ClassDescriptors.getSerializableClassDescriptor(ClassDescriptors.java:152)
        at org.jboss.marshalling.river.ClassDescriptors.getSerializableClassDescriptor(ClassDescriptors.java:146)
        at org.jboss.marshalling.river.ClassDescriptors.<clinit>(ClassDescriptors.java:101)
        ... 72 more
```

### How to Reproduce?

I have created a Docker Compose reproducer [GitHub project](https://github.com/slominskir/keycloak_cluster):


1. Grab project
```
git clone -b jdk17_cluster https://github.com/slominskir/keycloak_cluster
cd keycloak_cluster
```

2. Launch Compose
```
docker compose up
```

Watch the log output.

### Anything else?

_No response_

--- a/quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/CLITestExtension.java
+++ b/quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/CLITestExtension.java
@@ -17,15 +17,27 @@
 
 package org.keycloak.it.junit5.extension;
 
-import static org.keycloak.it.junit5.extension.DistributionTest.ReInstall.BEFORE_ALL;
-import static org.keycloak.it.junit5.extension.DistributionType.RAW;
-import static org.keycloak.quarkus.runtime.Environment.forceTestLaunchMode;
-import static org.keycloak.quarkus.runtime.cli.command.Main.CONFIG_FILE_LONG_NAME;
-import static org.keycloak.quarkus.runtime.cli.command.Main.CONFIG_FILE_SHORT_NAME;
+import io.quarkus.deployment.util.FileUtil;
+import io.quarkus.runtime.configuration.QuarkusConfigFactory;
+import io.quarkus.test.junit.QuarkusMainTestExtension;
+import io.quarkus.test.junit.main.Launch;
+import io.quarkus.test.junit.main.LaunchResult;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
+import org.keycloak.it.utils.KeycloakDistribution;
+import org.keycloak.it.utils.RawDistRootPath;
+import org.keycloak.it.utils.RawKeycloakDistribution;
+import org.keycloak.quarkus.runtime.Environment;
+import org.keycloak.quarkus.runtime.cli.command.Start;
+import org.keycloak.quarkus.runtime.cli.command.StartDev;
+import org.keycloak.quarkus.runtime.configuration.KeycloakPropertiesConfigSource;
+import org.keycloak.quarkus.runtime.configuration.test.TestConfigArgsConfigSource;
+import org.keycloak.quarkus.runtime.integration.QuarkusPlatform;
 
 import java.io.IOException;
 import java.lang.reflect.Method;
-import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashSet;
@@ -35,25 +47,11 @@ import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import io.quarkus.deployment.util.FileUtil;
-import io.quarkus.runtime.configuration.QuarkusConfigFactory;
-import org.junit.jupiter.api.extension.ExtensionContext;
-import org.junit.jupiter.api.extension.ParameterContext;
-import org.junit.jupiter.api.extension.ParameterResolutionException;
-import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
-import org.keycloak.it.utils.RawDistRootPath;
-import org.keycloak.it.utils.KeycloakDistribution;
-import org.keycloak.it.utils.RawKeycloakDistribution;
-import org.keycloak.quarkus.runtime.Environment;
-import org.keycloak.quarkus.runtime.cli.command.Start;
-import org.keycloak.quarkus.runtime.cli.command.StartDev;
-
-import io.quarkus.test.junit.QuarkusMainTestExtension;
-import io.quarkus.test.junit.main.Launch;
-import io.quarkus.test.junit.main.LaunchResult;
-import org.keycloak.quarkus.runtime.configuration.KeycloakPropertiesConfigSource;
-import org.keycloak.quarkus.runtime.configuration.test.TestConfigArgsConfigSource;
-import org.keycloak.quarkus.runtime.integration.QuarkusPlatform;
+import static org.keycloak.it.junit5.extension.DistributionTest.ReInstall.BEFORE_ALL;
+import static org.keycloak.it.junit5.extension.DistributionType.RAW;
+import static org.keycloak.quarkus.runtime.Environment.forceTestLaunchMode;
+import static org.keycloak.quarkus.runtime.cli.command.Main.CONFIG_FILE_LONG_NAME;
+import static org.keycloak.quarkus.runtime.cli.command.Main.CONFIG_FILE_SHORT_NAME;
 
 public class CLITestExtension extends QuarkusMainTestExtension {
 
@@ -102,6 +100,9 @@ public class CLITestExtension extends QuarkusMainTestExtension {
             copyTestProvider(context.getRequiredTestMethod().getAnnotation(TestProvider.class));
             onBeforeStartDistribution(context.getRequiredTestClass().getAnnotation(BeforeStartDistribution.class));
             onBeforeStartDistribution(context.getRequiredTestMethod().getAnnotation(BeforeStartDistribution.class));
+
+            configureEnvVars(context.getRequiredTestClass().getAnnotation(WithEnvVars.class));
+            configureEnvVars(context.getRequiredTestMethod().getAnnotation(WithEnvVars.class));
 
             if (launch != null) {
                 result = dist.run(Stream.concat(List.of(launch.value()).stream(), List.of(distConfig.defaultOptions()).stream()).collect(Collectors.toList()));
@@ -331,6 +332,16 @@ public class CLITestExtension extends QuarkusMainTestExtension {
         setProperty("quarkus.vault.devservices.enabled", Boolean.FALSE.toString());
         setProperty("quarkus.datasource.devservices.enabled", Boolean.TRUE.toString());
         setProperty("quarkus.devservices.enabled", Boolean.TRUE.toString());
+    }
+
+    private void configureEnvVars(WithEnvVars envVars) {
+        if (envVars == null) {
+            return;
+        }
+
+        for (int i=0; i<envVars.value().length; i=i+2) {
+            dist.setEnvVar(envVars.value()[i], envVars.value()[i+1]);
+        }
     }
 
     private void setProperty(String name, String value) {

--- a/quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/WithEnvVars.java
+++ b/quarkus/tests/integration/src/main/java/org/keycloak/it/junit5/extension/WithEnvVars.java
@@ -1,0 +1,37 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.it.junit5.extension;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * {@link WithEnvVars} is used to set environment variables before starting the server
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+public @interface WithEnvVars {
+
+    /**
+     * List of environment variables to be set as key-value pairs, e.g. {"JAVA_OPTS", -Xms64m", "DEBUG_PORT", "5005"}
+     */
+    String[] value() default {};
+
+}

--- a/testsuite/integration-arquillian/tests/base/pom.xml
+++ b/testsuite/integration-arquillian/tests/base/pom.xml
@@ -364,7 +364,7 @@
                                     <run>
                                         <shmSize>${docker.database.shmsize}</shmSize>
                                         <ports>
-                                            <port>${docker.database.port}</port>
+                                            <port>${docker.database.port}:${docker.database.port}</port>
                                         </ports>
                                         <env>
                                             <!-- MYSQL and MariaDB -->

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/cluster/UserFederationInvalidationClusterTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/cluster/UserFederationInvalidationClusterTest.java
@@ -1,0 +1,96 @@
+package org.keycloak.testsuite.cluster;
+
+import org.apache.commons.lang.RandomStringUtils;
+import org.junit.Before;
+import org.keycloak.admin.client.resource.ComponentResource;
+import org.keycloak.admin.client.resource.ComponentsResource;
+import org.keycloak.representations.idm.ComponentRepresentation;
+import org.keycloak.storage.UserStorageProvider;
+import org.keycloak.testsuite.admin.ApiUtil;
+import org.keycloak.testsuite.arquillian.ContainerInfo;
+import org.keycloak.testsuite.federation.DummyUserFederationProviderFactory;
+
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.core.Response;
+
+import static org.junit.Assert.assertNull;
+
+public class UserFederationInvalidationClusterTest extends AbstractInvalidationClusterTestWithTestRealm<ComponentRepresentation, ComponentResource> {
+
+    @Before
+    public void setExcludedComparisonFields() {
+    }
+
+    @Override
+    protected ComponentRepresentation createTestEntityRepresentation() {
+        ComponentRepresentation comp = new ComponentRepresentation();
+        comp.setName("comp_" + RandomStringUtils.randomAlphabetic(5));
+
+        // The provider needs to implement ImportSynchronization to trigger UserStorageSyncManager#notifyToRefreshPeriodicSync
+        comp.setProviderId(DummyUserFederationProviderFactory.PROVIDER_NAME);
+        comp.setProviderType(UserStorageProvider.class.getName());
+        return comp;
+    }
+
+    protected ComponentsResource components(ContainerInfo node) {
+        return getAdminClientFor(node).realm(testRealmName).components();
+    }
+
+    @Override
+    protected ComponentResource entityResource(ComponentRepresentation comp, ContainerInfo node) {
+        return entityResource(comp.getId(), node);
+    }
+
+    @Override
+    protected ComponentResource entityResource(String id, ContainerInfo node) {
+        return components(node).component(id);
+    }
+
+    @Override
+    protected ComponentRepresentation createEntity(ComponentRepresentation comp, ContainerInfo node) {
+        comp.setParentId(getAdminClientFor(node).realm(testRealmName).toRepresentation().getId());
+        try (Response response = components(node).add(comp)) {
+            String id = ApiUtil.getCreatedId(response);
+            comp.setId(id);
+        }
+        return readEntity(comp, node);
+    }
+
+    @Override
+    protected ComponentRepresentation readEntity(ComponentRepresentation comp, ContainerInfo node) {
+        ComponentRepresentation u = null;
+        try {
+            u = entityResource(comp, node).toRepresentation();
+        } catch (NotFoundException nfe) {
+            // expected when component doesn't exist
+        }
+        return u;
+    }
+
+    @Override
+    protected ComponentRepresentation updateEntity(ComponentRepresentation comp, ContainerInfo node) {
+        entityResource(comp, node).update(comp);
+        return readEntity(comp, node);
+    }
+
+    @Override
+    protected void deleteEntity(ComponentRepresentation comp, ContainerInfo node) {
+        entityResource(comp, node).remove();
+        assertNull(readEntity(comp, node));
+    }
+
+    @Override
+    protected ComponentRepresentation testEntityUpdates(ComponentRepresentation comp, boolean backendFailover) {
+        comp.setName(comp.getName() + "_updated");
+        comp = updateEntityOnCurrentFailNode(comp, "name");
+        verifyEntityUpdateDuringFailover(comp, backendFailover);
+
+        return comp;
+    }
+
+    @Override
+    protected void assertEntityOnSurvivorNodesEqualsTo(ComponentRepresentation testEntityOnFailNode) {
+        super.assertEntityOnSurvivorNodesEqualsTo(testEntityOnFailNode);
+    }
+
+}

--- a/testsuite/model/pom.xml
+++ b/testsuite/model/pom.xml
@@ -160,7 +160,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
-                    <argLine>@{argLine} -Xmx1024m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError</argLine>
+                    <!-- See also https://github.com/wildfly/wildfly-core/blob/7e5624cf92ebe4b64a4793a8c0b2a340c0d6d363/core-feature-pack/common/src/main/resources/content/bin/common.sh#L57-L60 -->
+                    <argLine>@{argLine} -Xmx1024m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED</argLine>
                     <systemPropertyVariables>
                         <!-- keycloak.model.parameters lists parameter classes from 
                              org.keycloak.model.parameters package and determine enabled providers -->

